Integrate UI with the Stats V4 Top Performers API

ThomazFB opened the pull request:
Summary
==========
Fixes issue woocommerce/woocommerce-admin#2538, this PR makes the entire `MyStoreFragment` compliant with the Stats V4 Top Performers API. Replacing the current usage of the `WCTopEarnerModel` of the `MyStoreDashboard` with the new `WCTopPerformerProductModel` as implemented on issue woocommerce/woocommerce-admin#2537. Also, this PR improves the usage of the offline capabilities of the app when retrieving the Top Performers data and introduces the usage of coroutines on the `MyStorePresenter`.

Changelog
==========
| Description | Commit |
| ------------- | ------------- |
| Migrates the current implementation of the Top Earners to the Top Performers from Stats V4 API | 7d832986b718ab213ce3e31abfaa363b44fbbe53 | 
| Fix the bug introduced on the last commit caused by changing the wrong fetch call | c6aca79bbc0ff113417adc415e38211823e35b17 | 
| Make the Top Performers call capable of using the forced call handling | 88cfe3ac63a20df22e1548e0f02440f2de363f70 | 
| Improve suspended functions definitions around the `MyStorePresenter` | 834368653305e0923c7359ddfa06ff4542d3bb5c | 
| Adapt the `MyStorePresenter` tests to work with the new infrastructure around coroutines | 380f32254b7878c5173fe912fb035ff8331d55d9 | 

Screenshots
==========
| When Top Performers data available | When there's no Top Performers data available |
| ------------- | ------------- |
| ![Screenshot_1596776373](https://user-images.githubusercontent.com/5920403/89611441-9e6c3c00-d853-11ea-988c-7273d87e4cc9.png) | ![Screenshot_1596776378](https://user-images.githubusercontent.com/5920403/89611443-a035ff80-d853-11ea-98da-8b88036fc958.png) |

How to Test
==========
1. Log in with a store running V4 Woo and *does have* WCAdmin installed.
2. On the My Store page make sure that the Stats V4 Top Performers is showing up and loading the stats synced with the Woo Site data.
3. Click on a Top Performer Product and verify if the Product Details View is open and works as it should.

Update release notes:

- [X] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

ThomazFB committed:
Create basic integration with WCLeaderboardsStore calls

ThomazFB committed:
Fix revenue retrieval issue

ThomazFB committed:
Update FluxC Leaderboards code checkout reference

ThomazFB committed:
Support forced option for Leaderboards requests

ThomazFB committed:
Adjust MyStoreContract Top Performers calls to become suspend functions

ThomazFB committed:
Fix MyStorePresenter tests

ThomazFB committed:
Refactor for lint, naming pattern and tests

ThomazFB committed:
Fix skeleton lifecycle control

loremattei commented:
Hey! I'm moving this to 4.9 since 4.8 has been cut. Since it's still a draft, I think it's fine. Otherwise, feel free to ping me to get this included in 4.8.

ThomazFB committed:
Merge branch 'develop' into feature/support-top-performers-v4

ThomazFB committed:
Change FluxC checkout version

AmandaRiu commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/mystore/MyStorePresenter.kt:
```diff
@@ -62,8 +66,8 @@ class MyStorePresenter @Inject constructor(
             }
-            for (i in 0 until topEarnersForceRefresh.size) {
-                topEarnersForceRefresh[i] = true
+            for (i in 0 until topPerformersForceRefresh.size) {
+                topPerformersForceRefresh[i] = true
```
NP: lint warning - statement can be changed to:
```
for (i in topPerformersForceRefresh.indices) {
    topPerformersForceRefresh[i] = true
}
```

ThomazFB replied:
Hey Amanda, thanks for the warning about that, I've fixed the lint warnings for this and for another additional point of the code right here: https://github.com/woocommerce/woocommerce-android/pull/2676/commits/2799e905750c44cbfabb9b0862ad3e5acd0cb3c6

AmandaRiu reviewed (changes requested):
@ThomazFB Nice work on this! Tested and results match my store dashboard 👍 Also tested offline mode and that works fine. Even though it is all working I did find a couple things for your review:

The app doesn’t match the web as far as prioritization. Is this something that can be fixed?

![Screen Shot 2020-08-17 at 6 26 03 PM](https://user-images.githubusercontent.com/5810477/90574864-01c96880-e188-11ea-886b-8637acd17421.png)



There is an `IllegalArgumentException` being thrown in logical when I try to fetch the data to populate the Top Performers view. It doesn’t appear to prevent it from populating, but it’s definitely a new error (I checked out `develop` and that error is not thrown). Here is the stack trace:

```
2020-08-17 18:23:27.561 21610-21610/com.woocommerce.android.dev E/WordPress-UTILS: Error finding valid currency symbol for currency code [$] in locale [en_US]
    java.lang.IllegalArgumentException: The input currency code is not 3-letter alphabetic code.
        at android.icu.util.Currency.getInstance(Currency.java:291)
        at java.util.Currency.getInstance(Currency.java:289)
        at org.wordpress.android.fluxc.utils.WCCurrencyUtils.getLocalizedCurrencySymbolForCode(WCCurrencyUtils.kt:45)
        at org.wordpress.android.fluxc.store.WooCommerceStore.formatCurrencyForDisplay(WooCommerceStore.kt:170)
        at com.woocommerce.android.util.CurrencyFormatter.formatCurrencyRounded(CurrencyFormatter.kt:74)
        at com.woocommerce.android.ui.mystore.MyStoreFragment$onActivityCreated$4.invoke(MyStoreFragment.kt:148)
        at com.woocommerce.android.ui.mystore.MyStoreFragment$onActivityCreated$4.invoke(MyStoreFragment.kt:37)
        at com.woocommerce.android.ui.mystore.MyStoreTopPerformersView$TopEarnersAdapter.onBindViewHolder(MyStoreTopPerformersView.kt:150)
        at com.woocommerce.android.ui.mystore.MyStoreTopPerformersView$TopEarnersAdapter.onBindViewHolder(MyStoreTopPerformersView.kt:114)
        at androidx.recyclerview.widget.RecyclerView$Adapter.onBindViewHolder(RecyclerView.java:7065)
        at androidx.recyclerview.widget.RecyclerView$Adapter.bindViewHolder(RecyclerView.java:7107)
        at androidx.recyclerview.widget.RecyclerView$Recycler.tryBindViewHolderByDeadline(RecyclerView.java:6012)
        at androidx.recyclerview.widget.RecyclerView$Recycler.tryGetViewHolderForPositionByDeadline(RecyclerView.java:6279)
        at androidx.recyclerview.widget.RecyclerView$Recycler.getViewForPosition(RecyclerView.java:6118)
        at androidx.recyclerview.widget.RecyclerView$Recycler.getViewForPosition(RecyclerView.java:6114)
        at androidx.recyclerview.widget.LinearLayoutManager$LayoutState.next(LinearLayoutManager.java:2303)
        at androidx.recyclerview.widget.LinearLayoutManager.layoutChunk(LinearLayoutManager.java:1627)
        at androidx.recyclerview.widget.LinearLayoutManager.fill(LinearLayoutManager.java:1587)
        at androidx.recyclerview.widget.LinearLayoutManager.onLayoutChildren(LinearLayoutManager.java:665)
        at androidx.recyclerview.widget.RecyclerView.dispatchLayoutStep2(RecyclerView.java:4134)
        at androidx.recyclerview.widget.RecyclerView.onMeasure(RecyclerView.java:3540)
        at android.view.View.measure(View.java:24530)
        at android.widget.RelativeLayout.measureChildHorizontal(RelativeLayout.java:735)
        at android.widget.RelativeLayout.onMeasure(RelativeLayout.java:481)
        at android.view.View.measure(View.java:24530)
        at android.view.ViewGroup.measureChildWithMargins(ViewGroup.java:6828)
        at android.widget.LinearLayout.measureChildBeforeLayout(LinearLayout.java:1552)
        at android.widget.LinearLayout.measureVertical(LinearLayout.java:842)
        at android.widget.LinearLayout.onMeasure(LinearLayout.java:721)
        at android.view.View.measure(View.java:24530)
        at android.view.ViewGroup.measureChildWithMargins(ViewGroup.java:6828)
        at android.widget.FrameLayout.onMeasure(FrameLayout.java:194)
        at androidx.cardview.widget.CardView.onMeasure(CardView.java:260)
        at com.google.android.material.card.MaterialCardView.onMeasure(MaterialCardView.java:159)
        at android.view.View.measure(View.java:24530)
        at android.view.ViewGroup.measureChildWithMargins(ViewGroup.java:6828)
        at android.widget.LinearLayout.measureChildBeforeLayout(LinearLayout.java:1552)
        at android.widget.LinearLayout.measureVertical(LinearLayout.java:842)
        at android.widget.LinearLayout.onMeasure(LinearLayout.java:721)
        at android.view.View.measure(View.java:24530)
        at androidx.core.widget.NestedScrollView.measureChildWithMargins(NestedScrollView.java:1599)
        at android.widget.FrameLayout.onMeasure(FrameLayout.java:194)
        at androidx.core.widget.NestedScrollView.onMeasure(NestedScrollView.java:585)
        at android.view.View.measure(View.java:24530)
        at androidx.swiperefreshlayout.widget.SwipeRefreshLayout.onMeasure(SwipeRefreshLayout.java:641)
        at android.view.View.measure(View.java:24530)
        at android.view.ViewGroup.measureChildWithMargins(ViewGroup.java:6828)
2020-08-17 18:23:27.561 21610-21610/com.woocommerce.android.dev E/WordPress-UTILS:     at android.widget.LinearLayout.measureChildBeforeLayout(LinearLayout.java:1552)
        at android.widget.LinearLayout.measureVertical(LinearLayout.java:842)
        at android.widget.LinearLayout.onMeasure(LinearLayout.java:721)
        at android.view.View.measure(View.java:24530)
        at android.view.ViewGroup.measureChildWithMargins(ViewGroup.java:6828)
        at android.widget.FrameLayout.onMeasure(FrameLayout.java:194)
        at android.view.View.measure(View.java:24530)
        at android.view.ViewGroup.measureChildWithMargins(ViewGroup.java:6828)
        at android.widget.FrameLayout.onMeasure(FrameLayout.java:194)
        at android.view.View.measure(View.java:24530)
        at android.view.ViewGroup.measureChildWithMargins(ViewGroup.java:6828)
        at androidx.coordinatorlayout.widget.CoordinatorLayout.onMeasureChild(CoordinatorLayout.java:760)
        at com.google.android.material.appbar.HeaderScrollingViewBehavior.onMeasureChild(HeaderScrollingViewBehavior.java:99)
        at com.google.android.material.appbar.AppBarLayout$ScrollingViewBehavior.onMeasureChild(AppBarLayout.java:1892)
        at androidx.coordinatorlayout.widget.CoordinatorLayout.onMeasure(CoordinatorLayout.java:831)
        at android.view.View.measure(View.java:24530)
        at androidx.constraintlayout.widget.ConstraintLayout.onMeasure(ConstraintLayout.java:1676)
        at android.view.View.measure(View.java:24530)
        at android.view.ViewGroup.measureChildWithMargins(ViewGroup.java:6828)
        at android.widget.FrameLayout.onMeasure(FrameLayout.java:194)
        at androidx.appcompat.widget.ContentFrameLayout.onMeasure(ContentFrameLayout.java:143)
        at android.view.View.measure(View.java:24530)
        at android.view.ViewGroup.measureChildWithMargins(ViewGroup.java:6828)
        at android.widget.LinearLayout.measureChildBeforeLayout(LinearLayout.java:1552)
        at android.widget.LinearLayout.measureVertical(LinearLayout.java:842)
        at android.widget.LinearLayout.onMeasure(LinearLayout.java:721)
        at android.view.View.measure(View.java:24530)
        at android.view.ViewGroup.measureChildWithMargins(ViewGroup.java:6828)
        at android.widget.FrameLayout.onMeasure(FrameLayout.java:194)
        at android.view.View.measure(View.java:24530)
        at android.view.ViewGroup.measureChildWithMargins(ViewGroup.java:6828)
        at android.widget.LinearLayout.measureChildBeforeLayout(LinearLayout.java:1552)
        at android.widget.LinearLayout.measureVertical(LinearLayout.java:842)
        at android.widget.LinearLayout.onMeasure(LinearLayout.java:721)
        at android.view.View.measure(View.java:24530)
        at android.view.ViewGroup.measureChildWithMargins(ViewGroup.java:6828)
        at android.widget.FrameLayout.onMeasure(FrameLayout.java:194)
        at com.android.internal.policy.DecorView.onMeasure(DecorView.java:742)
        at android.view.View.measure(View.java:24530)
        at android.view.ViewRootImpl.performMeasure(ViewRootImpl.java:3006)
        at android.view.ViewRootImpl.measureHierarchy(ViewRootImpl.java:1833)
        at android.view.ViewRootImpl.performTraversals(ViewRootImpl.java:2122)
        at android.view.ViewRootImpl.doTraversal(ViewRootImpl.java:1721)
        at android.view.ViewRootImpl$TraversalRunnable.run(ViewRootImpl.java:7598)
        at android.view.Choreographer$CallbackRecord.run(Choreographer.java:966)
        at android.view.Choreographer.doCallbacks(Choreographer.java:790)
        at android.view.Choreographer.doFrame(Choreographer.java:725)
        at android.view.Choreographer$FrameDisplayEventReceiver.run(Choreographer.java:951)
        at android.os.Handler.handleCallback(Handler.java:883)
        at android.os.Handler.dispatchMessage(Handler.java:100)
        at android.os.Looper.loop(Looper.java:214)
        at android.app.ActivityThread.main(ActivityThread.java:7356)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:492)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:930)
```

While doing some quick troubleshooting[I did find a bug that isn’t related to this one, but just so you’re aware of it here’s the link](https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/1669).

ThomazFB committed:
Merge branch 'develop' into feature/support-top-performers-v4

ThomazFB committed:
Adjust lint warnings

ThomazFB committed:
Add sorting to match site Leaderboards order

ThomazFB committed:
Fix currencyCode exception by using the Site Settings one

ThomazFB commented:
Hey @AmandaRiu, thanks for the review! I've added some changes and I think the code is ready for a second round 👍 

>The app doesn’t match the web as far as prioritization. Is this something that can be fixed?

Sure! I've added a presentation sorting rule right here: https://github.com/woocommerce/woocommerce-android/pull/2676/commits/2b257803728a2318ea47963706be3ef2506e2b79. The sorting rule I've added is inspired by the Woo site Leaderboards view, as shown below:

<img width="1728" alt="Screen Shot 2020-08-20 at 10 38 50 AM" src="https://user-images.githubusercontent.com/5920403/90792519-ef167700-e2e0-11ea-9ae9-6bcedf577282.png">

So, the main sorting considers as the sorting value the `total orders / items sold` and as a second sorting value the `total spend / net sales` as a diff rule for the first sort value. Here's the output on the app:

<img width="455" alt="Screen Shot 2020-08-20 at 12 35 03 PM" src="https://user-images.githubusercontent.com/5920403/90793042-9dbab780-e2e1-11ea-9bd4-0053a5042b8d.png">

>There is an IllegalArgumentException being thrown in logical when I try to fetch the data to populate the Top Performers view. It doesn’t appear to prevent it from populating, but it’s definitely a new error (I checked out develop and that error is not thrown). Here is the stack trace:

Thanks for detecting this one! As we talked on Slack, after a further investigation about this exception, I've found that the Stats V4 API doesn't return the currency code anymore but only the already formatted symbol in its place. At first, I thought this wasn't an issue since I saw the Woo app converting everything just fine, but after investigating this exception you pointed out, it looks like we can only know for sure that we can display the region currency format if we have the ISO code, I think the API didn't put that in consideration (especially considering that the currency and price are getting return as an HTML, which would already do the formatting job on a Woo site, but not on the app). I'll add this situation at woocommerce/woocommerce#32274, so we can track a final fix for that problem.

For now, I'm patching this using the currency code provided by the site settings, it should be safe, but isn't totally synced with the Top Performers data. Here's the change: https://github.com/woocommerce/woocommerce-android/pull/2676/commits/8ff6f198c57cdb28d902513b0a235a26f50d1260

It should solve the problem in a reliable way.

AmandaRiu reviewed (approved):
@ThomazFB The sorting is still off (see screenshot below), but that's really low priority so I'm not going to hold up this PR over it. Everything else is working great! :shipit: 

<img width="1097" alt="Screen Shot 2020-08-20 at 9 50 39 PM" src="https://user-images.githubusercontent.com/5810477/90843369-7559aa00-e32f-11ea-94d4-168c746f2169.png">
